File: worker/src/manifest.js
/* Pure manifest helpers shared by the /api/manifest routes.
 * No network, no platform APIs: given Scribd embed HTML (or a JSONP body),
 * extract the document manifest. Tested by manifest.test.js.
 */

const ORIG_RE =
  /orig=\\*"?(http:\/\/html\.scribd\.com\/[^"\\]+?\/images\/(\d+)-[a-f0-9]+\.(?:jpg|png))/gi;
const ADD_BLOCK_RE =
  /pageNum:\s*(\d+)\s*,.*?contentUrl:\s*"([^"]+)"/gs;
const COUNT_RE = /"page_count":\s*(\d+)/;
const TITLE_RE = /"title":\s*"([^"\\]{2,200})"/;
const DOC_RE =
  /^\/(?:document|presentation|doc|book|article|listen|embeds)\/\d+(?:\/|$)/;

export function toHttps(url) {
  return String(url).replace(
    "http://html.scribd.com/",
    "https://html.scribdassets.com/"
  );
}

export function extractDocId(input) {
  if (typeof input !== "string") return null;
  const value = input.trim();
  if (/^\d{5,}$/.test(value)) return value;
  try {
    const parsed = new URL(value);
    if (!["http:", "https:"].includes(parsed.protocol)) return null;
    const host = parsed.hostname.toLowerCase();
    if (host !== "scribd.com" && host !== "www.scribd.com") return null;
    if (!DOC_RE.test(parsed.pathname)) return null;
    return parsed.pathname.split("/")[2];
  } catch {
    return null;
  }
}

export function extractManifest(html) {
  const text = String(html ?? "");
  const count = COUNT_RE.exec(text);
  const title = TITLE_RE.exec(text);
  const inline = {};
  ORIG_RE.lastIndex = 0;
  let m;
  while ((m = ORIG_RE.exec(text))) {
    inline[Number(m[2])] = toHttps(m[1]);
  }
  const jsonpUrls = [];
  ADD_BLOCK_RE.lastIndex = 0;
  while ((m = ADD_BLOCK_RE.exec(text))) {
    jsonpUrls.push(m[2]);
  }
  return {
    title: title ? title[1] : "",
    pageCount: count ? Number(count[1]) : 0,
    inline,
    jsonpUrls,
  };
}

export function imageUrlFromJsonp(body) {
  ORIG_RE.lastIndex = 0;
  const m = ORIG_RE.exec(String(body ?? ""));
  return m ? toHttps(m[1]) : null;
}

File: relay/manifest.py
"""Pure manifest extraction from Scribd embed HTML.

No network, no browser: given the embed page HTML (fetched with a
Chrome-TLS-impersonating client from a residential IP), pull out the
document title, page count, inline page 1-3 image URLs and the ordered
per-page JSONP content URLs. JSONP bodies are parsed with the same
ORIG_RE (they escape quotes, which the pattern tolerates).
"""
import re

ORIG_RE = re.compile(
    r"orig=\\*\"?(http://html\.scribd\.com/[^\"\\]+?/images/(\d+)-[a-f0-9]+\.(?:jpg|png))",
    re.IGNORECASE,
)
ADD_BLOCK_RE = re.compile(r"pageNum:\s*(\d+)\s*,.*?contentUrl:\s*\"([^\"]+)\"", re.S)
COUNT_RE = re.compile(r"\"page_count\":\s*(\d+)")
TITLE_RE = re.compile(r"\"title\":\s*\"([^\"\\]{2,200})\"")


def to_https(url):
    return url.replace("http://html.scribd.com/", "https://html.scribdassets.com/")


def extract_manifest(html):
    """Return {title, page_count, inline:{page:url}, jsonp_urls:[...]}."""
    m = COUNT_RE.search(html)
    page_count = int(m.group(1)) if m else 0
    t = TITLE_RE.search(html)
    title = t.group(1) if t else ""
    inline = {}
    for mm in ORIG_RE.finditer(html):
        inline[int(mm.group(2))] = to_https(mm.group(1))
    jsonp_urls = [cu for _, cu in ADD_BLOCK_RE.findall(html)]
    return {
        "title": title,
        "page_count": page_count,
        "inline": inline,
        "jsonp_urls": jsonp_urls,
    }


def image_url_from_jsonp(body):
    """Pull the page image URL out of one JSONP body (jpg or png)."""
    m = ORIG_RE.search(body)
    return to_https(m.group(1)) if m else None
